plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+In Proceedings of the First ACM SIGPLAN International Conference on Functional Programming, 1-12.
+ICFP '96. New York, NY, USA: ACM, 1996. https://doi.org/10.1145/232627.232630.
+
+The algorithm of the paper deals with a different source language (Haskell/GHC),
+which only has 'letrec' and no support for letnonrec, type-lets, and datatype-lets.
+
+The paper's algorithm can be summarized to:
+
+* The first pass of the term annotates all let-bindings with their ranks (level numbers).
+The algorithm collects all the variables that occur free in a let's RHS and their ranks
+which were calculated before, at the place where those free variables were defined. The let
+is assigned the maximum of those ranks.
+* The second pass of the term uses those ranks "to float each binding outward to just
+outside the lambda which has a rank (level number) one greater than that on the binding".
+
+## Difference 1 - Rank
+
+For our case, a rank of a let is its closest-surrounding lambda "location" (depth + unique).
+In the paper, a rank of a let is calculated as the maximum among the ranks
+of the free-variable dependencies that occur in the let's RHS; a rank of a lambda is just its lambda-depth.
+The paper's rank includes only the lambda depth and not the lambda-unique.
+This is because the lambda-uniques are not needed by the original second pass because
+the re-positions of lets are applied "locally" (see Difference - Algorithm).
+
+## Difference - Algorithm
+
+The paper's algorithm applies the float-transformations "locally":
+during the second pass, when a let is encountered the algorithm will "decide" at that point to float it outwards or inwards.
+This allows more flexibility (heuristics and the possibility to float inwards).
+
+We use a different approach where we prior clean the AST PlutusIR term from all the lets (rec&nonrec)
+and store them in a separate 'RhsTable' (see `removeLets` function). We use the information obtained by the dependency graph
+to create let-(rec&nonrec) groups at the closest-surrounding lambda location (rank). During
+the second pass, whenever we reach a lambda/Lambda definition we float all lets belonging to
+this lambda rank (i.e. lambda's depth and unique).
+This means that the lets are floated *exactly under* a lambda definition, i.e at the lambda body.
+
+Contrast this with the paper's algorithm, where
+lets are floated exactly around the 'expr' that
+needs them, and *exactly outside a lambda definition*.
+This is because the algorithm stops re-positioning / floating the let-rhs, when it crosses over a lambda with depth
+one larger than its calculated rank (level number). Thus, for every lambda rank, a single merged let-group may be generated;
+this is no problem for Haskell/GHC, because all lets are letrecs.
+The advantage compared to our approach is that the lets will be floated closer to the 'expr' that needs them.
+Unfortunately, we cannot use this algorithm in our case for 2 reasons:  a) we cannot float past lambdas and b)
```

Comment:
It's worth to state explicitly why we can't float past lambdas and give an example.